src/commands/gen-api-models/render.ts

```
@@ -156,8 +166,12 @@ export const renderOperation = (
 
   const requestType = `r.I${capitalize(method)}ApiRequestType`;
 
+  const standardHeadersOnly = headers.filter(h => standardHeaders.includes(h));
```

If I get it right, this strips off all non-standard headers. My concerns are this is done silently, so the user will never know about it. Am I right?

Which makes me think about... do we need this concept of "standard headers"? What's that for?

---

You're right, but please notice that somehow it's already happening when the header is defined at method level.
You can see the [openapi](https://github.com/pagopa/openapi-codegen-ts/blob/ef454a7fa382ac1b492eb16356a69cc332128bbd/__mocks__/api.yaml#L134) and the related [snapshot](https://github.com/pagopa/openapi-codegen-ts/blob/ef454a7fa382ac1b492eb16356a69cc332128bbd/src/commands/gen-api-models/__tests__/__snapshots__/index.test.ts.snap#L366). 

> Which makes me think about... do we need this concept of "standard headers"? What's that for?

Standard headers are the ones defined in [@pagopa/ts-commons](https://github.com/pagopa/io-ts-commons/blob/ce618574bd0a847fa5571344f69072790ba12893/src/requests.ts#L25), but I don't know why this restriction was setup in the first place. 
Maybe we can think about upgrading that type to a more inclusive one.

---

What if we just accept any string, and let the "validation" to happen on the input specification? I mean: an invalid header should not be included in the spec in the first place. 

Maybe @gunzip remembers the need for such a constraint.

---

no as it was set up by @cloudify

---

@gquadrati to summarize, what are the changes expected from this PR? 

| When...  | Before      | After |
| - | - | - |
| ...a custom header is specified | Code is generated, but then fails to build | Code is generated but custom headers are removed silently |

---

Yes, and actually there is another change:

| When...  | Before      | After |
| - | - | - |
| ...an optional header is specified | Code is generated, but then fails to build | Code is generated  considering the header parameter as optional |

---

@gquadrati about custom headers: I see there will be no real-life changes to actual users, still we're producing a result which silently ignores users' input. It's a red-flag for me as it may lead to very unexpected issues, I'd rather fail the code generation.
Anyway I'm fine with the proposed changes, too, so we can merge if you think it's ok.

Let's discuss the need for restrict which header a user can define on another PR. I see no value nowadays, we may decide is a useless constraint.

---

I don't like the idea to make the generation fail as it's not a user problem, he has done everything right.
I don't like this solution neither, but I suppose changing `@pagopa/ts-commons` types will require a bit of attention.

We can propose the user to put headers as method parameter and not path param, and just fix the optional header issue.
@balanza  What do you think about it?

---

I think we can merge code as is and then we think about remove the constraint on standard headers.

Rationale: users should be able to provide any OpenAPI spec **as long as it's valid**, it's not this module's responsibility to argue on **what** the user puts in the spec.

---

fine for me.